### DISCUSSION TITLE
Documentation for dependencies

### DISCUSSION THREAD

**Discussion by Fogapod** [fastapi/fastapi#9106] [Questions]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * Or, I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```python
from fastapi import FastAPI, APIRouter, Depends, HTTPException, status

app = FastAPI()
router = APIRouter()

def get_token() -> str:
    raise HTTPException(status.HTTP_401_UNAUTHORIZED)

@router.get("/foo")
async def foo(token: str = Depends(get_token)):
    return {"text": "I can raise 401"}

@router.get("/bar")
async def bar():
    return {"text": "I do not raise anything"}

app.include_router(router)
```

### Description

Since dependencies can interfere request processing (raising exceptions), I think that it's an oversight that they can not contribute to generated documentation.

There are 2 possible solutions to document 401 response here:
* define  it in `APIRouter()`:
This is not sufficient because `/bar` will have it too even though it will never return this code.
* copypaste same responses everywhere where `get_token` is used:
This will work with 2 endpoints, but not with 200.

### The solution you would like

My suggestion is to add an **optional** dependency decorator with some subset of endpoint/router parameters to customize documetation:

```python
from fastapi import FastAPI, APIRouter, Depends, HTTPException, status
from fastapi import dependency  # new inport

app = FastAPI()
router = APIRouter()

@dependency(
    responses={status.HTTP_400_BAD_REQUEST: {"description": "Unauthorized"}}
)  # new decorator
def get_token() -> str:
    raise HTTPException(status.HTTP_401_UNAUTHORIZED)

@router.get("/foo")
async def foo(token: str = Depends(get_token)):
    return {"text": "I can raise 401"}

@router.get("/bar")
async def bar():
    return {"text": "I do not raise anything"}

app.include_router(router)
```

### Implementation details
* This can be implemented by adding some magic variable like `__fastapi_dependency_settings__` to function with decorator.
* If nested dependencies define conflicting status code (or endpoint itself define conflicting response status), top one should be used.
* This solution is fully backwards compatible. Not sure about `dependency` decorator name though.

**Comment by jarmani** [comment]
@tiangolo Any idea on this one ? Maybe another way to do it ?

Or if you are ok with the idea, I can try to implement it.

**Comment by lietu** [comment]
This is pretty much exactly the thing I'd want to see in FastAPI. The most common thing that annoys me with FastAPI is that to get responses in the OpenAPI documentation I need to add them manually to the responses, and there are no real options for how to make it happen automagically. I was considering some sort of metaprogramming options with decorators, trying to figure out what router the route is attached to and modifying it, but this would seem like a much more reasonable resolution. The main goal of documenting response types for dependencies is accomplished neatly here, and it would not really change the existing code while providing new optional functionality. This should of course be supported also on `@router.get("/foo", dependencies=[Depends(get_token)])`
